doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+or provide the relevant permissions for Elektra to work.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `malloc` calls primarily as no
+more memory could be allocated for the application. Such errors will gain special handling
+in future releases.
+
+##### General Resource ("C01120")
+
+`General Resource Errors` are all kind of permission and existence errors. Examples are
+missing files/ directories or insufficient permission to execute certain commands (eg. you
+would require sudo permissions). Reactions are fixing the permissions or creating the file/directory
+and retry the operation.
+
+#### Installation ("C01200")
+
+`Installation Errors` are those errors which are related to a wrong installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Installation errors might also be non-Elektra specific but also from dependent library/applications
+such as gpg.
+Also plugin configuration errors belong to `Installation Errors` as they this happens during
+mounting.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
+get rid of this error or fix the installation of the corresponding library/application.
+
+#### Logical ("C01300")
+
+`Logical Errors` is another branch category in which you indicate a bug in Elektra
+such as internal errors, assertion failures or errors
+which should not happen such as going into a `default` branch when you are assured that all cases
+are covered. Usually such errors come with a message to report such failures to Elektra's bugtracker.
+Applications cannot handle such errors themselves.
+
+##### Assertion ("C01310")
+
+`Assertion Errors` are such errors which indicate a flaw in the internal logic such as going into a `default`
+branch which you do never expect to happen.
+This category might be used in the future to automatically
+issue a bug to our bugtracker. As of now if you have to use this error please add a message
+indicating that this bug should be reported.
+
+##### Interface ("C01320")
+
+`Interface Errors` errors indicate a wrong usage of Elektra's API. An example would be to pass a NULL pointer to
+`kdbGet`. Also violations of the backend belong into this category. Compared to semantic validation errors,
+this category has its focus on detecting wrong usages of the API instead of a "retry with a different value" approach.
+
+##### Plugin is Broken ("C01330")
+
+`Broken Plugin Errors` errors indicate that a plugin does not behave in an intended way. Unrecognized commands,
+unkown return codes, plugin creation errors, etc. belong to this category. Also uncaught exceptions belong here because
+Elektra expects all exceptions to be caught.
+
+### Conflicting State ("C02000")
+
+`Conflicting State Errors` are errors where the current state is incompatible with the attempted operation.
```

Comment:
Explanation could be improved.

It should start with an explanation like: "Elektra has opportunistic locking...."

Reply:
Can you please help me in this regards? I have not that deep insight on how opportunistic locking in elektra works and why this can cause an error. Can you roughly explain it to me?